### ISSUE TITLE
Create a system to extend the expiration date of the login as soon as an alias from its group logs in

### ISSUE THREAD

**Issue by Madriix** [anope/anope#355]
Hi. 
For years, I've been using the sql_authentication module, and users log in either with their login or a group alias. I've noticed that sometimes, after 30-35 days, if they haven't logged in using their actual login because they only use their alias, the login expires and the alias becomes orphaned—it's no longer linked to any account, even when immediately recreated by the user. This issue stems from the sql_authentication module. The best solution I can think of to resolve this would be to extend the expiration date of the login whenever an alias logs in. I'm losing users when their login is deleted, rendering the alias unusable because it's no longer connected to any login since it has expired.

If this feature that extends the login timeout every time an alias logs in could be implemented, that would be great. 

Like for example in `anope.conf `(or `nickserv.conf`) there could be an option to configure:

> extend_date_login_on_alias_login = true

The login is the master nickname of the group

or 
> extend_date_master_nickname_on_alias_login = true

Regards

**Comment by TehPeGaSuS** [comment]
So said "master" nicks aren't extended even when you login to a grouped nick on IRC.
That's how it's supposed to be.
You extend the life of a nick you've logged in to.
Extending the "master" nick would only allow for nick hoarding.
IMHO, this behaviour should not be changed (unless I'm misunderstanding things here).
If you want to extend the life of a specific nick, log in to said nickname.

**Comment by Madriix** [comment]
` If you want to extend the life of a specific nick, log in to said nickname.`

I have already considered building a simple script either in node.js or in PHP, and the code would retrieve all master nicknames (but not the aliases). The retrieved nicknames would only be master nicknames that have at least one alias in the group. Thus, the script would detect the master nicknames that will expire in less than 5 days, and if an alias exists and the identification has been recently made on it, then the robot or the script would create a system to authenticate with it, for example with `/ns identify <nick_master> <password_md5>` - The `m_sql_authentication` module system takes into account the MD5. However, creating such a script would take time; it would be simpler to advance the expiration date of the master nickname on Anope as soon as an alias authenticates.

When a master nickname expires, the alias or aliases become orphaned and no longer correspond to the master nickname in the sql_authentication table. The aliases are not recognized; there are no aliases in `sql_authentication`. Each time, I have to manually delete all the aliases and ask users to identify with the master nickname and then recreate the aliases (or the alias if there is only one)

**Comment by Madriix** [comment]
Ah thanks, I will test. It won't be easy to test right now. I'll just have to wait and see if someone I know reports this issue to me, maybe in about a month. Some people let things slide without notifying me. Is there a way in the anope.db to modify the timestamp of one of my test nicknames so that it appears to be 30 or 29 days old? This would help me check if it expires tomorrow as soon as the alias exists and has been recently identified.

**Comment by Madriix** [comment]
@SadieCat I have this error:

```
[ 91%] Building CXX object modules/CMakeFiles/sql_authentication.so.dir/sql_authentication.cpp.o
/home/irc/anope2/anope-2.1.2/modules/sql_authentication.cpp: In member function ‘virtual void ModuleSQLAuthentication::OnPreNickExpire(NickAlias*, bool&)’:
/home/irc/anope2/anope-2.1.2/modules/sql_authentication.cpp:156:68: error: ‘class Serialize::Checker<std::vector<NickAlias*> >’ has no member named ‘size’
  156 |                 if (na->nick == na->nc->display && na->nc->aliases.size() > 1 && disable_reason.empty())
      |                                                                    ^~~~
make[2]: *** [modules/CMakeFiles/sql_authentication.so.dir/build.make:76 : modules/CMakeFiles/sql_authentication.so.dir/sql_authentication.cpp.o] Erreur 1
make[1]: *** [CMakeFiles/Makefile2:4573 : modules/CMakeFiles/sql_authentication.so.dir/all] Erreur 2
make: *** [Makefile:156 : all] Erreur 2
```

The Anope version I have is 2.1.2

**Comment by Madriix** [comment]
I am a novice in cpp, I replaced:
`na->nc->aliases.size()`
by
`na->nc->aliases->size()`

I also replaced:
` anope_override`
by
` override`

	void OnPreNickExpire(NickAlias *na, bool &expire) override
	{
		// We can't let nicks expire if they still have a group or
		// there will be a zombie account left over that can't be
		// authenticated to.
		if (na->nick == na->nc->display && na->nc->aliases->size() > 1 && disable_reason.empty())
			expire = false;
	}

The anope it started.

**Comment by SadieCat** [comment]
Oops yeah the change I linked is for the 2.0 branch. I'll merge it across to 2.1 when it gets committed though.

**Comment by Madriix** [comment]
I modified the timestamp of the last_seen in anope.db for a test pseudonym. So it expires tonight in 2 hours for this test master pseudonym. I also created an alias in the Madrix group, which was identified today. I will wait tonight or tomorrow to see if the master pseudonym has been deleted.

Is it indeed the timestamp of the last_seen that needs to be modified? Is there anything else to modify elsewhere?

> DATA time_registered 1684267209
> DATA **last_seen 1703179177**
> DATA nc Madrix
